File: 128 Longest Consecutive Sequence/128longest-consecutive-sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0
        nums_set = set(nums)
        res = 1

        for num in nums_set:
            if num - 1 in nums_set:
                continue
            curr_seq = 1
            while num + 1 in nums_set:
                curr_seq += 1
                num += 1
            res = max(res, curr_seq)
        return res

# Time: O(n + n)
# Space: O(n)

File: 128 Longest Consecutive Sequence/128longest-consecutive-sequence.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> numSet(nums.begin(), nums.end());
        int res = 0;

        for (int num: numSet) {
            if (numSet.find(num - 1) == numSet.end()) {
                int curr_len = 1;
                while (numSet.find(num + 1) != numSet.end()) {
                    curr_len += 1;
                    num++;
                }
                res = max(res, curr_len);
            }
        }
        return res;
    }
};

// Time: O(n + n)
// Space: O(n)